yt/utilities/lib/cyoctree.pyx

```diff
@@ -0,0 +1,590 @@
+    cdef int _max_depth
+    cdef int _over_refine_factor    # this allows the tree to be built with more
+                                    # than 8 cells per leaf
+    cdef int _num_cells             # 2**(3*_over_refine_factor)
+    cdef int _num_cells_per_dim     # 2**(_over_refine_factor)
+
+    # Children structure
+    cdef int _density_factor        # this allows the tree to be built with more
+                                    # than 8 children per node
+    cdef int _num_children          # 2**(3*_density_factor)
+    cdef int _num_children_per_dim  # 2**(_density_factor)
+
+    # this is use for interpolation and is global for the octree smoothing
+    # operations
+    cdef kernel_func kernel
+
+    def __init__(self, double[:, ::1] &input_pos = None, left_edge = None,
+                 right_edge = None, int n_ref=32, int over_refine_factor=1,
+                 int density_factor=1, np.int64_t data_version=0,
+                 int max_depth=20):
+
+        # if this is the case, we are very likely just initialising an instance
+        # and then going to load an existing Octree from memory, so we don't
+        # really need to do anything
+        if input_pos is None:
+            return
+
+        # These don't have setters as these would invalidate the tree which is
+        # a feature we don't have
+        self._data_version = data_version
+        self._n_ref = n_ref
+        self._num_particles = input_pos.shape[0]
+        self._max_depth = max_depth
+
+        # setting the properties which determines how children divide
+        self.over_refine_factor = over_refine_factor
+        self.density_factor = density_factor
+
+        # set up the initial idx of the particles, this keeps track of which
+        # particle from the input is in which node
+        self._idx = np.arange(0, input_pos.shape[0], dtype=np.int64)
+
+        # set up the bounds and root node
+        self.setup_bounds(input_pos, left_edge, right_edge)
+        self.setup_root(input_pos)
+
+        # reserve some space for the octs to be stored
+        reserve(&self.nodes, 1000000)
+
+        # now build the tree
+        self.build_tree(&input_pos[0, 0])
+
+    cdef setup_bounds(self, double[:, ::1] &input_pos, left_edge=None,
+                      right_edge=None):
+        if left_edge is not None:
+            for i in range(3):
+                self._left_edge[i] = left_edge[i]
+        else:
+            for i in range(3):
+                self._left_edge[i] = np.amin(input_pos[:,i])
+
+        if right_edge is not None:
+            for i in range(3):
+                self._right_edge[i] = right_edge[i]
+        else:
+            for i in range(3):
+                self._right_edge[i] = np.amax(input_pos[:,i])
+
+    cdef setup_root(self, double[:, ::1] &input_pos):
+        cdef Node root
+        root.left_edge = self._left_edge
+        root.right_edge = self._right_edge
+        root.parent = 0
+
+        root.start = 0
```

Comment:
we assume all particles given are in the octree bounds, in the case of the yt data object I *think* this is always true, but I can add some stuff here to accommodate cases where this isnt true